### PR TITLE
Introduce Runnable

### PR DESCRIPTION
**Pull request by ttung**
Runnable is the base unit of a recipe.  Each runnable constitutes a PipelineComponent with a specific algorithm.  Constructor arguments that are FileProviders are loaded into memory and passed in.  Inputs to the `run()` method can also include FileProviders.  In all cases, FileProviders are associated with a file path or url.  The type is inferred from the typing parameters of the method (currently supported: ImageStack, IntensityTable, ExpressionMatrix, and Codebook).

Runnables will be wired together to constitute a pipeline recipe.

Test plan: Added tests to verify a simple Runnable, chained Runnables, and Runnables that has constructor arguments that are FileProviders.

This is a continuation of #1097

### CODE REVIEW

File: starfish/recipe/filesystem.py [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+    associated with the implementing class, the method to invoke to load such a filetype, and the
+    method to invoke to save back to the filetype.
+
+    The load method is expected to be called with a string, which is the file or url to load from,
+    and is expected to return an instantiated object.
+
+    The save method is expected to be called with the object and a string, which is the path to
+    write the object to.
+    """
+    IMAGESTACK = (ImageStack, imagestack_convert, ImageStack.export)
```
Comment (ambrosejcarr) [inline starfish/recipe/filesystem.py]: Scope creep: I feel the save/load names for our methods should be consistent across the 4 objects we have. What do you think?

File: starfish/recipe/filesystem.py [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+    @staticmethod
+    def resolve_by_instance(instance) -> "FileTypes":
+        for member in FileTypes.__members__.values():
+            if isinstance(instance, member.value[0]):
+                return member
+        raise TypeError(f"filetype of {instance.__class__} not supported.")
+
+
+class FileProvider:
+    """This is used to wrap paths or URLs that are passed into Runnables via the `file_inputs` magic
```
Comment (ambrosejcarr) [inline starfish/recipe/filesystem.py]: `file_inputs` is not used anywhere in this PR, just mentioned in docs. Confused.

File: starfish/recipe/filesystem.py (hunk 1 above)
Comment (ttung) [reply starfish/recipe/filesystem.py]: https://github.com/spacetx/starfish/pull/1097#discussion_r267532690

LOL you and @shanaxel42 think alike. :)

Comment (ttung) [reply starfish/recipe/filesystem.py]: https://github.com/spacetx/starfish/issues/1183

File: starfish/recipe/runnable.py [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+    RunInsufficientParametersError,
+    TypeInferenceError,
+)
+from .filesystem import FileProvider, TypedFileProvider
+
+
+class Runnable:
+    """Runnable represents a single invocation of a pipeline component, with a specific algorithm
+    implementation.  For arguments to the algorithm's constructor and run method, it can accept
+    :py:class:`starfish.recipe.filesystem.FileProvider` objects, which represent a file path or url.
```
Comment (shanaxel42) [inline starfish/recipe/runnable.py]: doc tip! Adding ~ before the classpath renders the link as just `FileProvider` so ```:py:class:`~starfish.recipe.filesystem.FileProvider```

File: starfish/recipe/runnable.py [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+            A callable that can be invoked to provide a user-friendly representation of the
+            Runnable, in case any errors or warnings are generated.
+
+        Returns
+        -------
+        Mapping[str, Any] : The parameters for the constructor, ready to be passed into the
+        constructor.
+        """
+        parameters = constructor_signature.parameters
+        assert next(iter(parameters.keys())) == "self"
```
Comment (shanaxel42) [inline starfish/recipe/runnable.py]: why are you checking this?

File: starfish/recipe/test/test_runnable.py [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+    assert str(filter_runnable) == ("compute(\"filter\", \"SimpleFilterAlgorithm\","
+                                    + f" FileProvider(\"{URL}\"), multiplicand=0.5)")
+
+
+def test_constructor_error():
+    """Verify that we get a properly typed error when the constructor does not execute correctly."""
+    with pytest.raises(ConstructorError):
+        Runnable(
+            "filter", "SimpleFilterAlgorithm",
+            FileProvider(URL),
```
Comment (shanaxel42) [inline starfish/recipe/test/test_runnable.py]: why does this error? because we don't provide params?

File: starfish/recipe/test/test_runnable.py [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+    with pytest.raises(TypeInferenceError):
+        Runnable(
+            "filter", "FilterAlgorithmWithMissingConstructorTyping",
+            FileProvider(URL),
+            additive=FileProvider(URL),
+        )
+
+
+def test_run_type_inference_error():
+    """Verify that we get a properly typed error when we cannot properly infer the type for one of
```
Comment (shanaxel42) [inline starfish/recipe/test/test_runnable.py]: Could use more explanation here too, maybe some comments about what input causes the error

File: starfish/recipe/runnable.py (hunk 3 above)
Comment (ttung) [reply starfish/recipe/runnable.py]: Done.

File: starfish/recipe/runnable.py (hunk 4 above)
Comment (ttung) [reply starfish/recipe/runnable.py]: I removed this.  I am keeping the equivalent sanity check for `run` in case we change the way we structure AlgorithmBase fundamentally (e.g., `run` becomes a static or class method).

File: starfish/recipe/test/test_runnable.py (hunk 6 above)
Comment (ttung) [reply starfish/recipe/test/test_runnable.py]: Fixed, and added more commentary on how to fix this error in the documentation for `TypeInferenceError`.

File: starfish/recipe/test/test_runnable.py (hunk 5 above)
Comment (ttung) [reply starfish/recipe/test/test_runnable.py]: Yes.  Updated the documentation.

**Comment by ttung** [comment]
Restacked on top of #1218

File: starfish/recipe/filesystem.py [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+from starfish.util.indirectfile import (
+    convert,
+    GetCodebook,
+    GetCodebookFromExperiment,
+    GetImageStack,
+    GetImageStackFromExperiment,
+)
+
+
+def imagestack_convert(indirect_path_or_url: str) -> ImageStack:
```
Comment (ambrosejcarr) [inline starfish/recipe/filesystem.py]: single-line docstring please: what is this intended to do?

File: starfish/recipe/filesystem.py [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+    return convert(
+        indirect_path_or_url,
+        [
+            GetImageStack(),
+            GetImageStackFromExperiment(),
+        ],
+    )
+
+
+def codebook_convert(indirect_path_or_url: str) -> Codebook:
```
Comment (ambrosejcarr) [inline starfish/recipe/filesystem.py]: single-line docstring please: what is this intended to do?

File: starfish/recipe/filesystem.py (hunk 7 above)
Comment (ttung) [reply starfish/recipe/filesystem.py]: Added doc.

File: starfish/recipe/filesystem.py (hunk 8 above)
Comment (ttung) [reply starfish/recipe/filesystem.py]: Added doc.
